fhir-cql-server/src/main/java/com/ibm/fhir/cql/engine/server/retrieve/ServerFhirRetrieveProvider.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * (C) Copyright IBM Corp. 2021
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package com.ibm.fhir.cql.engine.server.retrieve;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+
+import javax.ws.rs.core.MultivaluedHashMap;
+import javax.ws.rs.core.MultivaluedMap;
+
+import com.ibm.fhir.cql.engine.retrieve.SearchParameterFhirRetrieveProvider;
+import com.ibm.fhir.cql.engine.retrieve.SearchParameterMap;
+import com.ibm.fhir.cql.engine.searchparam.IQueryParameter;
+import com.ibm.fhir.cql.engine.searchparam.ReferenceParameter;
+import com.ibm.fhir.cql.engine.searchparam.SearchParameterResolver;
+import com.ibm.fhir.cql.helpers.FhirBundleCursor;
+import com.ibm.fhir.model.resource.Bundle;
+import com.ibm.fhir.model.resource.Resource;
+import com.ibm.fhir.persistence.SingleResourceResult;
+import com.ibm.fhir.search.SearchConstants;
+import com.ibm.fhir.server.operation.spi.FHIRResourceHelpers;
+
+public class ServerFhirRetrieveProvider extends SearchParameterFhirRetrieveProvider {
+
+    private static final String DUMMY_REQUEST_URI = "https://localhost/fhir-server/api/v4";
+
+    private static Logger logger = Logger.getLogger(ServerFhirRetrieveProvider.class.getName());
+    
+    private FHIRResourceHelpers resourceHelpers;
+
+    public ServerFhirRetrieveProvider(FHIRResourceHelpers resourceHelpers, SearchParameterResolver searchParameterResolver) {
+        super(searchParameterResolver);
+        this.resourceHelpers = resourceHelpers;
+    }
+
+    @Override
+    protected Iterable<Object> executeQueries(String dataType, List<SearchParameterMap> queries) throws Exception {
+        List<Object> results = new ArrayList<>();
+
+        for (SearchParameterMap map : queries) {
+            if( logger.isLoggable(Level.FINE) ) {
+                logger.fine(String.format("Executing query %s", map.toString()));
+            }
+            
+            MultivaluedMap<String, String> queryParameters = getQueryParameters(map);
+
+            // _total=none instructs the server to skip the count(*) query which improves performance
+            queryParameters.putSingle(SearchConstants.TOTAL, "none");
+
+            Resource resource = executeQuery(dataType, queryParameters);
+            if (resource instanceof Bundle) {
+                AtomicInteger pageNumber = new AtomicInteger(1);
+                FhirBundleCursor cursor = new FhirBundleCursor(url -> {
+                    try {
+                        int nextPage = pageNumber.incrementAndGet();
```

Comment:
increment and get? can you tell me what happens when you have two queries? 
is increment and get on the second query 2? 3 on second query? Shouldn't it start with page 1 and do a loop in the doSearch block?

Reply:
The cursor is instantiated from a Bundle resource that has already been retrieved (see line 58). That initial bundle is page 1. If, after reading all the entries in the bundle, we find a "next" link, then we update the query to request the next page which is 2.